pkg/dockerregistry/server/digestcache.go

```diff
@@ -23,23 +27,42 @@ func newDigestToRepositoryCache(size int) (digestToRepositoryCache, error) {
 	if err != nil {
 		return digestToRepositoryCache{}, err
 	}
-	return digestToRepositoryCache{Cache: c}, nil
+	return digestToRepositoryCache{
+		Cache: c,
+		clock: &util.RealClock{},
+	}, nil
 }
 
-const bucketSize = 10
+const bucketSize = 16
 
 // RememberDigest associates a digest with a repository.
-func (c digestToRepositoryCache) RememberDigest(dgst digest.Digest, repo string) {
+func (c digestToRepositoryCache) RememberDigest(dgst digest.Digest, ttl time.Duration, repo string) {
 	key := dgst.String()
 	value, ok := c.Get(key)
 	if !ok {
-		value = &repositoryBucket{}
-		if ok, _ := c.ContainsOrAdd(key, value); !ok {
-			return
+		value = &repositoryBucket{clock: c.clock}
+		if ok, _ := c.ContainsOrAdd(key, value); ok {
+			// the value exists now, get it
+			value, ok = c.Get(key)
+			if !ok {
+				// should not happen
+				return
+			}
 		}
 	}
 	repos := value.(*repositoryBucket)
-	repos.Add(repo)
+	repos.Add(ttl, repo)
+}
+
+// ForgetDigest removes an association between given digest and repository from the cache.
+func (c digestToRepositoryCache) ForgetDigest(dgst digest.Digest, repo string) {
+	key := dgst.String()
+	value, ok := c.Get(key)
```

Comment:
not sure we want to bump to the front of the LRU cache when removing a repo... maybe use `Peek()` instead?